tiledb/sm/filter/filter.cc

```diff
@@ -106,28 +117,38 @@ Status Filter::set_option(FilterOption option, const void* value) {
   return set_option_impl(option, value);
 }
 
-Status Filter::deserialize(ConstBuffer* buff, Filter** filter) {
+std::tuple<Status, std::optional<std::shared_ptr<Filter>>> Filter::deserialize(
+    ConstBuffer* buff) {
+  Status st;
   uint8_t type;
-  RETURN_NOT_OK(buff->read(&type, sizeof(uint8_t)));
+  st = buff->read(&type, sizeof(uint8_t));
+  if (!st.ok()) {
+    return {st, std::nullopt};
+  }
   uint32_t filter_metadata_len;
-  RETURN_NOT_OK(buff->read(&filter_metadata_len, sizeof(uint32_t)));
+  st = buff->read(&filter_metadata_len, sizeof(uint32_t));
+  if (!st.ok()) {
+    return {st, std::nullopt};
+  }
 
-  auto* f = create(static_cast<FilterType>(type));
-  if (f == nullptr)
-    return LOG_STATUS(Status::FilterError("Deserialization error."));
+  auto&& [st_filter, filter]{create(static_cast<FilterType>(type))};
```

Comment:
The `create` function you're calling isn't C.41 compliant, because it relies on later calls as a second-phase initialization. In this case the `deserialize_impl` calls below are second-phase calls.

Because we're not altering the way the C API works in the present PR, we'll need two internal functions like `create`. The existing one can be used as-is for the C API. We'll need another one for `deserialize`. The new one need not be separate from `deserialize`, as it would only be called once from within it.